File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """      
        n=len(candidates)
        res=[]
        candidates.sort()
        self.combo(0,target,[],candidates,res)
        return res 
    
    
    def combo(self,i,target,ds,arr,res):
        if target==0:
            res.append(ds)
            return

        for j in xrange(i,len(arr)):
            if j>i and arr[j]==arr[j-1]:
                continue
            if arr[j]>target:
                break
            self.combo(j+1,target-arr[j],ds+[arr[j]],arr,res)
            

File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        
        List<List<Integer>> result = new ArrayList<>();
        helper(0, candidates, new ArrayList<>(), result, target);
        return result;
    }
    
    private void helper(int index, int[] candidates, List<Integer> path, List<List<Integer>> result, int target){
        // Base Case
        if (target == 0){
            result.add(new ArrayList<>(path));
            return;
        }
        if ((target < 0) || (index >= candidates.length)){
            return;
        }
        
        for (int i = index; i < candidates.length; i++){
            if ((i > index) && (candidates[i] == candidates[i - 1])) continue;
            path.add(candidates[i]);
            helper(i + 1, candidates, path, result, target - candidates[i]);
            path.remove(path.size() - 1);
        }
    }
}